qiskit/pulse/transforms/canonicalization.py

```diff
@@ -85,24 +85,31 @@ def compress_pulses(schedules: List[Schedule]) -> List[Schedule]:
     Returns:
         Compressed schedules.
     """
-    existing_pulses = []
+    pulse_index = 0
+    existing_pulse_set = set()
+    pulse_index_map = dict()    # mapping pulse -> index
```

Comment:
I think the keys of `pulse_index_map` are the same set as `existing_pulse_set` here?  If so, you can just do the `in` checks on `pulse_index_map` and get rid of `existing_pulse_set`.